### ISSUE TITLE
Support workload identity federation for config connector

### ISSUE THREAD

**Issue by liubog2008** [GoogleCloudPlatform/k8s-config-connector#629]
**Describe the feature or resource**

Now it can only use static key.json to setup config connector. I hope to use workload identity federation to authenticate GCP API.

What I hope is deployed config connector in an external Kubernetes cluster and use it as an OIDC provider to exchange GCP token and then use this token to visit GCP API. 

**Importance**

Links: 
- https://cloud.google.com/iam/docs/using-workload-identity-federation#oidc_2
- https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection

**Comment by liubog2008** [comment]
It has been supported in cnrm-controller-manager in fact. However, I have to modify sts after the controller is deployed by operator.

> And would you consider this a blocker, friction point, or nice-to-have?

I feel it's a blocker for me to use config connector in prod unless I choose to deploy controller manually.

> it's a feature that lets one use GKE Workload Identity on non-GKE clusters?

Yes, and we also use it in GKE cluster. We use k8s cluster as an OIDC provider to authenticate GCP API. Pod can use the token which is signed by k8s to exchange a new token signed by GCP to visit GCP API.


There are some changes of cnrm-controller-manager. I'm not sure whether it is all what I need to change. Just for reference only.

cnrm-controller-manager sts.yaml
```
        env:
        - name: GOOGLE_APPLICATION_CREDENTIALS
          value: /var/secrets/google/key.json
        volumeMounts:
        - mountPath: /var/secrets/google
          name: gcp-service-account
        - mountPath: /var/run/secrets/gke
          name: gke-token
      volumes:
      - name: gke-token
        projected:
          defaultMode: 420
          sources:
          - serviceAccountToken:
              audience: test.com
              expirationSeconds: 7200
              path: token
      - name: gcp-service-account
        secret:
          defaultMode: 420
          secretName: gsa-key
```

key.json

```
{
  "type": "external_account",
  "audience": "//iam.googleapis.com/projects/{project_number}/locations/global/workloadIdentityPools/{pool}/providers/{provider}",
  "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
  "token_url": "https://sts.googleapis.com/v1/token",
  "credential_source": {
    "file": "/var/run/secrets/gke/token"
  },
  "service_account_impersonation_url": "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/test@{project_id}.iam.gserviceaccount.com:generateAccessToken"
}
```

**Comment by jcanseco** [comment]
Gotcha, thanks for elaborating @liubog2008. The examples are particularly helpful.

I agree that this sound useful. I'm afraid I don't see us being able to work on this any time soon though due to capacity constraints -- I just want to make sure those expectations are set clearly so you're not waiting on a feature that we can't deliver in a reasonable timeframe.

That said, we'll let you know when that changes.

To other customers: please also like this issue if you think this helps your use-case. It'll help us prioritize accordingly later.

**Comment by liubog2008** [comment]
Fine, I'll try to deploy controller manually.

**Comment by feisley** [comment]
@liubog2008 did you ever find a way to get this working with Config Connector. We are looking to do the exact same thing with an external Kubernetes clusted that is federated with Google via Workload Identity Federation. We need that external cluster to be able to create GCP resources and would like to avoid putting a static service account key into that cluster when it already has the ability to use federated credentials.

@jcanseco we are also very interested in this capability and this is critical for us to be able to do so securely in production.

**Comment by liubog2008** [comment]
@feisley just see my above example. You can use an external kubernetes cluster as an OIDC provider.

**Comment by feisley** [comment]
@liubog2008 ok, on my first read it looked you suggested that as a possible way it could be implemented but after looking at it it will just work as-is.

We simply use the clientConfig.json that workload identity federation produces and place that into config connector and it should set it to the 'GOOGLE_APPLICATION_CREDENTIALS' environment variable which in turn all the rest of the Google SDK should recognize and auto-auth.

**Comment by ldanielmadariaga** [comment]
This item has been open for >12 months. As part of scheduled repository maintenance, this Issue/PR will be closed. If this is still a critical ask, please re-open and work with your Google Cloud Account team to prioritize it with GCP KCC Product Management.
